src/storage/src/controller/rehydration.rs

```diff
@@ -183,6 +185,39 @@ where
                 .await;
         }
 
+        for export in self.exports.values_mut() {
+            let mut persist_clients = self.persist.lock().await;
+            let persist_client = persist_clients
+                .open(
+                    export
+                        .description
+                        .from_storage_metadata
+                        .persist_location
+                        .clone(),
+                )
+                .await
+                .expect("error creating persist client");
+            let from_read_handle = persist_client
+                .open_reader::<SourceData, (), T, Diff>(
+                    export.description.from_storage_metadata.data_shard,
+                )
+                .await
+                .expect("from collection disappeared");
+
+            let cached_as_of = &export.description.as_of.frontier;
+            // The controller has the dependency recorded in it's `exported_collections` so this
+            // should not change at least until the sink is started up (because the storage
+            // controller will not downgrade the source's since).
+            let from_since = from_read_handle.since();
```

Comment:
Now that I look at this I'm a bit concerned about races. The following might happen:

1. controller holds a read handle for the implied since `t`
2. the actual since of the shard is `t - 10` because someone else is still holding a handle
3. we set the `as_of` to `t - 10` because of this
4. that third party released their hold, shard since advances to `t`
5. the sink, in `storaged` tries to read and fails

I think 1. and 2. can't currently happen together, because we know that the controller initializes the implied frontier to the shard since. But we might change that in the future, maybe by accident.

It would be nicer if we can thread through the exact since that the controller is holding, but that might not be easily feasible. In that case we should probably merge as is. Tricky ... 🙈

Reply:
I think the correctness here is slightly more embedded than you describe: one of the first things the storage controller does is install a dependency between the `from` source / table / etc and this newly created sink.  That will keep it from updating the read capability of the source

so, while someone else can definitely downgrade their handle, the handle managed by the storage controller should keep the since for the collection itself from being downgraded